rust_src/src/keymap.rs

```diff
@@ -73,6 +73,59 @@ pub fn keymap_parent_lisp(keymap: LispObject) -> LispObject {
     LispObject::from_raw(keymap_parent(keymap.to_raw(), true))
 }
 
+/// Check whether MAP is one of MAPS parents.
+#[no_mangle]
+pub extern "C" fn keymap_memberp(map: Lisp_Object, maps: Lisp_Object) -> bool {
+    let map = LispObject::from_raw(map);
+    let mut maps = LispObject::from_raw(maps);
+    if map.is_nil() {
+        return false;
+    }
+    while keymapp(maps) && map != maps {
+        maps = LispObject::from_raw(keymap_parent(maps.to_raw(), false));
+    }
+    map == maps
+}
+
+/// Modify KEYMAP to set its parent map to PARENT.
+/// Return PARENT.  PARENT should be nil or another keymap.
+#[lisp_fn]
+pub fn set_keymap_parent(keymap: LispObject, parent: LispObject) -> LispObject {
+    let mut list;
+    let mut prev;
+    let mut parent = parent;
+
+    // Flush any reverse-map cache
+    unsafe {
+        where_is_cache = Qnil;
+        where_is_cache_keymaps = Qt;
+    }
+
+    let keymap = unsafe { LispObject::from_raw(get_keymap(keymap.to_raw(), true, true)) };
+    if parent.is_not_nil() {
+        parent = unsafe { LispObject::from_raw(get_keymap(parent.to_raw(), true, false)) };
+
+        // Check for cycles
+        if keymap_memberp(keymap.to_raw(), parent.to_raw()) {
+            error!("Cyclic keymap inheritance");
+        }
+    }
+
+    // Skip past the initial element 'keymap'.
+    prev = keymap.as_cons_or_error();
+    loop {
+        list = prev.cdr();
+        // If there is a parent keymap here, replace it.
+        // If we came to the end, add the parent in PREV.
+        if !list.is_cons() || keymapp(list) {
+            prev.check_impure();
+            prev.set_cdr(parent);
+            return parent;
+        }
+        prev = list.as_cons_or_error();
+    }
+}
+
```

Comment:
How about this instead.

        loop {
            list = prev.cdr();

            match list.as_cons() {
                None => break,
                Some(cons) => if keymapp(list) {
                    break;
                } else {
                    prev = cons;
                }
            }
        }
        prev.check_impure();
        prev.set_cdr(parent);
        return parent;

Reply:
ok, maybe no 'return' :-)

Reply:
The code you had checked `is_cons` twice. Once with explicit `is_cons` and once with `as_cons_or_error`. The original code did not have that error. Instead it implicitly claimed that prev would always be a `Lisp_Cons`. My code makes this explicit and removes the double check. If there is another way to write this cleanly I am open to it.

In our opinion one of the reasons the Rust port has value is it removes the need for the assertions because we can know what types are involved. At least some of the time. This chunk is a demonstration of why we do not usually need `XCDR`.

Reply:
I get what you mean. I'll be revisiting this later to fully so I can fully comprehend. Thanks!

Reply:
BTW, If the return lives in the `loop` that is ok. I only pulled it out to simplify the code I wrote.